src/tests/suite/dfuse_test.c

```diff
@@ -176,25 +189,154 @@ do_ioctl(void **state)
 	assert_int_equal(dur.uid, geteuid());
 	assert_int_equal(dur.gid, getegid());
 
+	rc = unlinkat(root, "ioctl_file", 0);
+	assert_return_code(rc, errno);
+
 	rc = close(root);
 	assert_return_code(rc, errno);
 }
 
-int
-main(int argc, char **argv)
+static bool
+timespec_gt(struct timespec t1, struct timespec t2)
+{
+	if (t1.tv_sec == t2.tv_sec)
+		return t1.tv_nsec > t2.tv_nsec;
+	else
+		return t1.tv_sec > t2.tv_sec;
+}
+
+void
+do_mtime(void **state)
+{
+	struct stat     stbuf;
+	struct timespec prev_ts;
+	struct timespec now;
+	struct timespec times[2];
+	int             fd;
+	int             rc;
+	char            input_buf[] = "hello";
+	int             root        = open(test_dir, O_PATH | O_DIRECTORY);
+
+	assert_return_code(root, errno);
+
+	/* Open a file and sanity check the mtime */
+	fd = openat(root, "my_file", O_RDWR | O_CREAT | O_EXCL, S_IWUSR | S_IRUSR);
+	assert_return_code(fd, errno);
+	rc = clock_gettime(CLOCK_REALTIME, &now);
+	assert_return_code(fd, errno);
+	rc = fstat(fd, &stbuf);
+	assert_return_code(rc, errno);
+	prev_ts.tv_sec  = stbuf.st_mtim.tv_sec;
+	prev_ts.tv_nsec = stbuf.st_mtim.tv_nsec;
+	assert_true(now.tv_sec - prev_ts.tv_sec < 3);
+
+	/* Write to the file and verify mtime is newer */
+	rc = write(fd, input_buf, sizeof(input_buf));
+	assert_return_code(rc, errno);
+	rc = fstat(fd, &stbuf);
+	assert_return_code(rc, errno);
+	assert_true(timespec_gt(stbuf.st_mtim, prev_ts));
+	prev_ts.tv_sec  = stbuf.st_mtim.tv_sec;
+	prev_ts.tv_nsec = stbuf.st_mtim.tv_nsec;
+
+	/* Truncate the file and verify mtime is newer */
+	rc = ftruncate(fd, 0);
+	assert_return_code(rc, errno);
+	rc = fstat(fd, &stbuf);
+	assert_return_code(rc, errno);
+	assert_true(timespec_gt(stbuf.st_mtim, prev_ts));
+	prev_ts.tv_sec  = stbuf.st_mtim.tv_sec;
+	prev_ts.tv_nsec = stbuf.st_mtim.tv_nsec;
+
+	/* Set and verify mtime set in the past */
+	times[0]         = now;
+	times[1].tv_sec  = now.tv_sec - 10;
```

Comment:
i understand that 10 seconds is probably way more than enough time for the entire test to run, however in CI and especially with VM tests, we've seen some pretty weird behavior and some tests taking very long without any reason. so maybe this past time step should use the first timestamp on the file as a base instead of now.

Reply:
@ashleypittman This is my doing, but I agree with Mohamad. It's more robust to use the first timestamp (or prev?) instead of `now`.

Reply:
We've got further work to do with this feature that is blocked on this so I'm going to land as-is and we can review if it becomes a problem.